Add test cases for Option<&T> and fix rust codegen

PrismaPhonic opened the pull request:
Add test cases for Option<&T> and fix rust codegen

Currently swift-bridge only has tests for Option<T> - this adds test
cases for Option<&T> and fixes a bug in rust codegen that does not
correctly translate Option<&T>.

This is now possible:

```rust
mod ffi {
  extern "Rust" {
    type MyStruct;

    fn my_func(arg: Option<&MyStruct>) -> Option<&MyStruct>;
  }
}
```

chinedufn commented on crates/swift-bridge-ir/src/codegen/codegen_tests/option_codegen_tests.rs:
```diff
@@ -460,6 +460,64 @@ void* __swift_bridge__$some_function(void);
+    fn expected_swift_code() -> ExpectedSwiftCode {
+        ExpectedSwiftCode::ContainsAfterTrim(
+            r#"
+func some_function() -> Optional<SomeType> {
```
This should be `Optional<SomeTypeRef>`

PrismaPhonic replied:
Done

chinedufn commented on crates/swift-bridge-ir/src/codegen/codegen_tests/option_codegen_tests.rs:
```diff
@@ -460,6 +460,64 @@ void* __swift_bridge__$some_function(void);
+        ExpectedSwiftCode::ContainsAfterTrim(
+            r#"
+func some_function() -> Optional<SomeType> {
+    { let val = __swift_bridge__$some_function(); if val != nil { return SomeType(ptr: val!) } else { return nil } }()
```
`SomeType` -> `SomeTypeRef`

chinedufn commented on SwiftRustIntegrationTestRunner/SwiftRustIntegrationTestRunnerTests/OptionTests.swift:
```diff
@@ -97,6 +97,14 @@ class OptionTests: XCTestCase {
+
+    func testSwiftCallRustWithOptionRefOpaqueRustType() throws {
+        let val = OptTestOpaqueRustTypeRef(123)
+        let reflect = rust_reflect_option_ref_opaque_rust_type(val)
```
I think that this might be UB at the moment, but should be easy to fix:

Let's add an extra call to `rust_reflect_option_ref_opaque_rust_type` so that we can be confident that we are returning a reference.

By calling it twice we can gain a bit more confidence that we aren't trying to free the reference.

So something along the lines of:

```swift
/// Verify that we can pass and receive an `Option<&RustType>`.
///
/// We deinitialize the first reference and create a second to confirm that
/// deinitializing the reference does not deinitialize the Rust type.
func testSwiftCallRustWithOptionRefOpaqueRustType() throws {
    let val = OptTestOpaqueRustTypeRef(123)

    let reflect = rust_reflect_option_ref_opaque_rust_type(val)
    XCTAssertEqual(reflect!.field(), 123)
    XCTAssertNil(rust_reflect_option_ref_opaque_rust_type(nil))
    let reflect = nil;

    let second_reference = rust_reflect_option_ref_opaque_rust_type(val);
    XCTAssertEqual(second_reference!.field(), 123)
}
```

---

From what I gather this test should fail. I'll comment on how to make it pass below.

PrismaPhonic replied:
Updated so we check it twice

chinedufn reviewed (commented):
Looks great. Thanks a lot for taking the time to put this together. 

Looks like you found the `if self.reference { quote! { *const super::#type_name #generics } }` bit. Awesome.

---

Just two changes please:

---

Avoid dropping the underlying Rust type when the Swift side drops the reference to the Rust type.
This is explained in the PR comments.

---

Update your PR body to show an example of a bridge module that did not work before this commit but works as of this commit.

Example: https://github.com/chinedufn/swift-bridge/pull/203#issue-1630545202

````
This commit adds support for Option<PrimitiveType> in `extern "Swift"`
functions.

For example, the following is now possible:

```rust
mod ffi {
    extern "Swift" {
        fn my_func(arg: Option<u8>) -> Option<bool>;
    }
}
```
````

---

After that this looks great and can land. Thank you for putting this together.

chinedufn commented on crates/swift-integration-tests/src/option.rs:
```diff
@@ -178,6 +191,20 @@ impl OptTestOpaqueRustType {
+
+impl OptTestOpaqueRustTypeRef {
+    fn new(field: u8) -> &'static Self {
+        &Self { field, }
```
This won't compile since we're trying to return a reference to a temporary value.

Can do something like:

```rust
pub struct OptTestOpaqueRustTypeRef {
    field: u8,
}

impl OptTestOpaqueRustTypeRef {
    fn new_some(field: u8) -> Option<Self> {
        Some(Self { field })
    }

    fn field(&self) -> u8 {
        self.field
    }
}

fn rust_convert_option_ty_to_option_ref_ty(val: Option<OptTestOpaqueRustTypeRef>) -> Option<&OptTestOpaqueRustTypeRef> {
    val.as_ref()
}
```

Then on the Swift side can call `OptTestOpaqueRustTypeRef.new_some` and then `rust_convert_option_ty_to_option_ref_ty` to get the `Option<&OptTestOpaqueRustTypeRef>` that you need.

---

Here's how to call an associated method such as `new_some` https://github.com/chinedufn/swift-bridge/blob/master/book/src/bridge-module/functions/README.md?plain=1#L146-L189

chinedufn replied:
Ah whoops.

This is one solution:

```rust
fn rust_convert_ref_ty_to_option_ref_ty(val: &OptTestOpaqueRustTypeRef) -> Option<&OptTestOpaqueRustTypeRef> {
    Some(val)
}
``

PrismaPhonic replied:
Maybe something like this?

```rust
pub struct OptTestOpaqueRustTypeRef {
    field: Option<OptTestOpaqueRustType>,
}

impl OptTestOpaqueRustTypeRef {
    fn new(field: u8) -> Self {
        Some(Self { field: Some(OptTestOpaqueRustType::new(field)) })
    }

    fn field(&self) -> Option<OptTestOpaqueRustType> {
        self.field
    }

    fn field_ref(&self) -> &Option<OptTestOpaqueRustType> {
        &self.field
    }
}

fn rust_convert_ref_option_ty_to_option_ref_ty(
    val: &Option<OptTestOpaqueRustType>,
) -> Option<&OptTestOpaqueRustType> {
    val.as_ref()
}
```

chinedufn replied:
Only challenge is that I'm not sure if `&Option<T>` is supported yet. SO the `val: &Option<OptTestOpaqueRustType>` might not work.

I like the general direction though. What about:

```rust
pub struct OptTestOpaqueRustTypeRef {
    field: Option<OptTestOpaqueRustType>,
}

impl OptTestOpaqueRustTypeRef {
    fn new(field: u8) -> Self {
        Some(Self { field: Some(OptTestOpaqueRustType::new(field)) })
    }

    fn field_ref(&self) -> Option<&OptTestOpaqueRustType> {
        self.field.as_ref()
    }
}
```

Then the test can call `.field_ref()` and then reflect the `Option<&OptTestOpaqueRustType>`.

PrismaPhonic replied:
Oh right - that should work! Because then the parent object still owns the inner type. I'll try that :) thanks!

chinedufn replied:
Team work made the dream work!

PrismaPhonic replied:
Alright, got that pushed up - hitting another compiler error (that I hit before) listed below in a general comment

chinedufn commented on crates/swift-integration-tests/src/option.rs:
```diff
@@ -41,6 +41,15 @@ mod ffi {
+    extern "Rust" {
+        type OptTestOpaqueRustTypeRef;
+
+        #[swift_bridge(init)]
```
Can remove this `#[swift_bridge(init)]` as per the feedback below.

PrismaPhonic replied:
Done

chinedufn reviewed (commented):
Left some notes on how to pass CI.

Also need to run `cargo fmt`.

PrismaPhonic commented on crates/swift-bridge-ir/src/codegen/codegen_tests/option_codegen_tests.rs:
```diff
@@ -460,6 +460,64 @@ void* __swift_bridge__$some_function(void);
     }
 }
 
+/// Test code generation for Rust function that returns an Option<&OpaqueRustType>
```
Hahahaha, thank you. This is my first foray into FFI land so I appreciate it. 

I think I am missing what is needed here. Another `bridge_module_tokens()` method?

chinedufn replied:
Gotcha:

1. Copy the module `extern_rust_fn_return_option_ref_opaque_rust_type` to a new module below it `extern_rust_fn_arg_option_ref_opaque_rust_type`

2. Replace the bridge module in `extern_rust_fn_arg_option_ref_opaque_rust_type` with the one in the comment above.

3. `extern_rust_fn_arg_option_ref_opaque_rust_type` will pass

4. Tweak the `expected_rust_tokens` to make it pass

PrismaPhonic replied:
Alright - did that to the best of my abilities and pushed up.

PrismaPhonic commented:
I've addressed all concerns listed here but seem to still hit this compiler error:

```rust
error[E0308]: mismatched types
   --> crates/swift-integration-tests/src/option.rs:95:13
    |
3   | #[swift_bridge::bridge]
    | ----------------------- arguments to this function are incorrect
...
95  |             arg: Option<&OptTestOpaqueRustType>,
    |             ^^^ types differ in mutability
    |
    = note: expected raw pointer `*mut _`
               found raw pointer `*const OptTestOpaqueRustType`
note: associated function defined here
   --> /home/pmfarr/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/boxed.rs:950:19
    |
950 |     pub unsafe fn from_raw(raw: *mut T) -> Self {
```

chinedufn commented:
> Do you know if there's way to have the compiler show the error at the expanded code?

I think there's a `rustc` flag for it but I've never used it https://stackoverflow.com/a/68843349

> I think that we are generating a function that takes in a *const (which it should) but then tries to call Box::from_raw() on it which wants a *mut - but I'm not sure where to fix that

Yup. Can do another `if self.reference` in here:

https://github.com/chinedufn/swift-bridge/blob/4fbd30f81085dc366bda3c0303eac66345de3ed9/crates/swift-bridge-ir/src/bridged_type/bridged_opaque_type.rs#L438

Something like:

```rust
if self.reference {
    Some(unsafe {& * #expression})
} else {
    Some(unsafe { * Box::from_raw(#expression) } )
}
```

chinedufn commented on crates/swift-bridge-ir/src/codegen/codegen_tests/option_codegen_tests.rs:
```diff
@@ -460,6 +460,122 @@ void* __swift_bridge__$some_function(void);
+            mod ffi {
+                extern "Rust" {
+                    type SomeType;
+                    fn some_function (arg: Option<&SomeType>) -> Option<&SomeType>;
```
I think this comment got swallowed since I was commenting on orphaned commits, so let me try here in the PR's "files changed" page.

Can remove the `-> Option<&SomeType>` here.

I made a mistake when I said to add it. The return case is already tested in the other test that you added so we don't need it here.

PrismaPhonic replied:
So remove the entire module then? or...

chinedufn replied:
Keep the module but remove the `-> Option<&SomeType>` from this line 530 right here

So ` fn some_function (arg: Option<&SomeType>) -> Option<&SomeType>;`

becomes ` fn some_function (arg: Option<&SomeType>);

chinedufn commented on crates/swift-bridge-ir/src/codegen/codegen_tests/option_codegen_tests.rs:
```diff
@@ -460,6 +460,122 @@ void* __swift_bridge__$some_function(void);
+    }
+}
+
+/// Test code generation for Rust function that returns an Option<&OpaqueRustType>
+mod extern_rust_fn_arg_option_ref_opaque_rust_type {
```
Just FYI here's a good test to take inspiration from: https://github.com/chinedufn/swift-bridge/blob/fdca9fcc2080a60d30de958fc4ada2516691581e/crates/swift-bridge-ir/src/codegen/codegen_tests/option_codegen_tests.rs#L463-L523

I'll comment below with the changes though, so no need to study that one too carefully.

chinedufn commented on crates/swift-bridge-ir/src/codegen/codegen_tests/option_codegen_tests.rs:
```diff
@@ -460,6 +460,122 @@ void* __swift_bridge__$some_function(void);
+
+    fn expected_rust_tokens() -> ExpectedRustTokens {
+        ExpectedRustTokens::Contains(quote! {
+            #[export_name = "__swift_bridge__$some_function"]
+            pub extern "C" fn __swift_bridge__some_function() -> *const super::SomeType {
+                if let Some(val) = super::some_function() {
+                    val as *const super::SomeType
+                } else {
+                    std::ptr::null()
+                }
+            }
```
This can be:

```rust
#[export_name = "__swift_bridge__$some_function"]
pub extern "C" fn __swift_bridge__some_function(
    arg: *const super::SomeType
) {
    super::some_function(
        if arg.is_null() {
            None
        } else {
            Some( unsafe { & * arg } )
        }
    )
}
```

chinedufn commented on crates/swift-bridge-ir/src/codegen/codegen_tests/option_codegen_tests.rs:
```diff
@@ -460,6 +460,122 @@ void* __swift_bridge__$some_function(void);
+    fn expected_swift_code() -> ExpectedSwiftCode {
+        ExpectedSwiftCode::ContainsAfterTrim(
+            r#"
+func some_function(arg: Optional<SomeTypeRef>) -> Optional<SomeTypeRef> {
+    { let val = __swift_bridge__$some_function(); if val != nil { return SomeTypeRef(ptr: val!) } else { return nil } }()
+}
```
This one is similar to: https://github.com/chinedufn/swift-bridge/blob/607e30579ac4de1fab5ff66838253cf3d07ddfbb/crates/swift-bridge-ir/src/codegen/codegen_tests/extern_rust_function_opaque_rust_type_argument_codegen_tests.rs#L91-L99

So it can be:

```swift
func some_function(_ arg: Optional<SomeTypeRef>) {
    __swift_bridge__$some_function({ if let val = arg { return val.ptr } else { return nil } }())
}
```

PrismaPhonic replied:
Seems that generated is this:

```swift
func some_function(_ arg: Optional<SomeTypeRef>) {
    __swift_bridge__$some_function({ if let val = arg { val.isOwned = false; return val.ptr } else { return nil } }())
}

```

Just checking that the assignment to setting isOwned to false is correct?

PrismaPhonic replied:
It does by the way look like codegen on the previous test (that you told me to change to `return SomeTypeRef>` is actually `return SomeType` - I'm guessing this is a mistake and something I need to correct in translation to swift somewhere. Do you know where that would be?

chinedufn replied:
`isOwned` is available on owned `SomeType` https://github.com/chinedufn/swift-bridge/blob/dd5bef56af28db4f1a1244d86115def6abede931/crates/swift-bridge-ir/src/codegen/codegen_tests/opaque_rust_type_codegen_tests.rs#L36-L48

but not on `SomeTypeRef` https://github.com/chinedufn/swift-bridge/blob/dd5bef56af28db4f1a1244d86115def6abede931/crates/swift-bridge-ir/src/codegen/codegen_tests/opaque_rust_type_codegen_tests.rs#L54-L60

---

So, we just need what should hopefully be one last `if self.reference` here https://github.com/chinedufn/swift-bridge/blob/4fbd30f81085dc366bda3c0303eac66345de3ed9/crates/swift-bridge-ir/src/bridged_type/bridged_opaque_type.rs#L312-L325

PrismaPhonic replied:
Oh wait - that type has access to know if it's a reference - in that case I can just modify `swift_name` to check `if self.reference` and append the suffix accordingly

chinedufn replied:
Hahaha look what I was about to comment with before I saw these last two comments:

https://github.com/chinedufn/swift-bridge/blob/4fbd30f81085dc366bda3c0303eac66345de3ed9/crates/swift-bridge-ir/src/bridged_type/bridged_opaque_type.rs#L639-L642

```rust
pub fn swift_name(&self) -> String {
	if self.reference {
		format!("{}", self.ty)
	} else {
		format!("{}Ref", self.ty)
	}
}
```

Your FFI-fu has grown strong my friend.

chinedufn replied:
I think that as you start handling more cases the two languages start to diverge but based on what we currently have tested they're always the same.

As in, you end up starting to add certain if statements based on the language.

So what happened here was probably something like:

1. if rust do X, if swift `todo!()`
2. one day we needed to add support for Swift so we wrote a test case
3. then we changed the logic to -> if rust do X, if swift do X
4. FUTURE -> if rust do X or Y depending on Z, if swift do X or Q depending on W

So, this is just in the middle of that transition to step 4.

I suppose you could collapse then if you wanted, but I think it's better as is so that when someone someday needs to modify one language they can clearly see where to do it (I haven't had to do this in a while so I forget what can cause the divergence).

If you disagree and think we should collapse back down until that day comes, feel free to collapse it.

PrismaPhonic replied:
One more spot I needed to check reference, but all working now!

In `convert_option_swift_expression_to_ffi_type`

chinedufn reviewed (commented):
I think this should do the trick

PrismaPhonic commented:
> > Do you know if there's way to have the compiler show the error at the expanded code?
> 
> I think there's a `rustc` flag for it but I've never used it https://stackoverflow.com/a/68843349
> 
> > I think that we are generating a function that takes in a *const (which it should) but then tries to call Box::from_raw() on it which wants a *mut - but I'm not sure where to fix that
> 
> Yup. Can do another `if self.reference` in here:
> 
> https://github.com/chinedufn/swift-bridge/blob/4fbd30f81085dc366bda3c0303eac66345de3ed9/crates/swift-bridge-ir/src/bridged_type/bridged_opaque_type.rs#L438
> 
> Something like:
> 
> ```rust
> if self.reference {
>     Some(unsafe {& * #expression})
> } else {
>     Some(unsafe { * Box::from_raw(#expression) } )
> }
> ```

Done, annnnnd now I'm hitting way more compiler errors :( 

```
error[E0424]: expected value, found module `self`
  --> crates/swift-integration-tests/src/option.rs:3:1
   |
3  | #[swift_bridge::bridge]
   | ^^^^^^^^^^^^^^^^^^^^^^^ `self` value is a keyword only available in methods with a `self` parameter
...
90 |         fn rust_reflect_option_opaque_rust_type(
   |            ------------------------------------ this function can't have a `self` parameter
   |
   = note: this error originates in the attribute macro `swift_bridge::bridge` (in Nightly builds, run with -Z macro-backtrace for more info)
```

Somehow now we are generating functions that don't have self to have self. I'm soooooo confused

chinedufn commented:
> Done, annnnnd now I'm hitting way more compiler errors :(

Ah gotcha. Just need the `self` to be outside of the `quote!` :

```rust
if self.reference {
    quote! {
        if #expression.is_null() {
            None
        } else {
            Some(unsafe {& * #expression})
        }
    }
} else {
    quote! {
        if #expression.is_null() {
            None
        } else {
            Some(unsafe { * Box::from_raw(#expression) } )
        }
    }
}
```

PrismaPhonic commented:
@chinedufn addressed all issues - all tests are passing for me locally (using the command in the README)

chinedufn commented on crates/swift-bridge-ir/src/bridged_type/.bridged_opaque_type.rs.rustfmt:
```diff
@@ -0,0 +1,764 @@
+use crate::bridged_type::{
```
Need to delete this file

PrismaPhonic replied:
oh shoot! Sorry about that, thought I avoided committing that one. Done

chinedufn commented on SwiftRustIntegrationTestRunner/SwiftRustIntegrationTestRunnerTests/OptionTests.swift:
```diff
@@ -97,6 +97,19 @@ class OptionTests: XCTestCase {
         XCTAssertNil(rust_reflect_option_opaque_rust_type(nil))
     }
+
+    func testSwiftCallRustWithOptionRefOpaqueRustType() throws {
```
Mind plopping this on here:

```rust
/// Verify that we can pass and receive an `Option<&RustType>`.
///
/// We deinitialize the first reference and create a second to confirm that
/// deinitializing the reference does not deinitialize the Rust type.
```

We'll eventually document every test in this way, so I want to enforce it for new ones.

The idea is that a new contributor can read the test docs and have a good sense of whether or not the test relates to what they're trying to do.

I was going to just add it myself but then I saw the `.rs.rustfmt` file so while commenting on that I figured I'd ask for this as well.

PrismaPhonic replied:
Good idea, done!

PrismaPhonic commented:
Huh, I wonder why integration tests failed in the pipeline but that one test in particular succeeds locally for me

chinedufn commented on crates/swift-integration-tests/src/option.rs:
```diff
@@ -35,10 +35,16 @@ mod ffi {
 
     extern "Rust" {
         type OptTestOpaqueRustType;
+        type OptTestOpaqueRustTypeRef;
```
Need to rename this. the `type OptTestOpaqueRustType` above generates `class OptTestOpaqueRustTypeRef` on the Swift side, and then this also generates a class with the same name.

<img width="1025" alt="image" src="https://github.com/chinedufn/swift-bridge/assets/2099811/20dee030-95cf-40c5-a83e-27bf297454db">

Can call it `OptTestOpaqueRefRustType` instead

PrismaPhonic replied:
ohhhhh right - that makes sense. Doh! magical strings hahaha

chinedufn replied:
Here's more information as to why this error is happening. 

No need to read this link to solve the problem.
Renaming the type (I gave an example under the screenshot in the previous comment) will do.

https://github.com/chinedufn/swift-bridge/blob/dd5bef56af28db4f1a1244d86115def6abede931/book/src/bridge-module/opaque-types/README.md?plain=1#L52-L133

PrismaPhonic replied:
Done and pushed up

PrismaPhonic replied:
Hmm, actually, now I'm getting this failure locally:

```
Testing failed:
	Missing argument label 'ptr:' in call
	Cannot convert value of type 'Int' to expected argument type 'UnsafeMutableRawPointer'
	'nil' requires a contextual type
	Command SwiftCompile failed with a nonzero exit code
	Testing cancelled because the build failed.
	```

PrismaPhonic replied:
I'm taking a wild guess here as I don't know swift that well but I'm guessing it's related to my logic here:

```swift
    func testSwiftCallRustWithOptionRefOpaqueRustType() throws {
        let val = OptTestOpaqueRefRustType.new(123)
        let opt_ref = val.field_ref()

        let reflect = rust_reflect_option_ref_opaque_rust_type(opt_ref)
        XCTAssertEqual(reflect!.field(), 123)
        XCTAssertNil(rust_reflect_option_ref_opaque_rust_type(nil))
        let reflect = nil
        
        let second_reference = rust_reflect_option_ref_opaque_rust_type(opt_ref)
        XCTAssertEqual(second_reference!.field(), 123)
    }
```

I wonder if this is the problem:

```swift
        XCTAssertEqual(reflect!.field(), 123)
```

Is it saying that the result of field() is a unsafe mutable raw pointer and it's trying to coerce the int of 123 into that but failing for the equality comparison?

In this case I call `val.field_ref()` which should get us the inner Option<&Type> and then I thought `reflect!.field()` goes inside that option and calls the `field()` method on the inner type which should then get us an int out?

chinedufn reviewed (approved):
Just need to delete a file and rename a type and then looks good to me.

Thanks for putting this together!``

chinedufn commented on SwiftRustIntegrationTestRunner/SwiftRustIntegrationTestRunnerTests/OptionTests.swift:
```diff
@@ -97,6 +97,23 @@ class OptionTests: XCTestCase {
+        let reflect = rust_reflect_option_ref_opaque_rust_type(opt_ref)
+        XCTAssertEqual(reflect!.field(), 123)
+        XCTAssertNil(rust_reflect_option_ref_opaque_rust_type(nil))
+        let reflect = nil
```
```swift
        var reflect = rust_reflect_option_ref_opaque_rust_type(opt_ref)
        XCTAssertEqual(reflect!.field(), 123)
        XCTAssertNil(rust_reflect_option_ref_opaque_rust_type(nil))
        reflect = nil
```

this way we can overwrite `reflect` and free the reference.

PrismaPhonic replied:
Done

chinedufn commented:
Mind pushing new commits instead of force pushing? This way I can see what has changed. I'll squash them all at the end.

chinedufn commented on crates/swift-integration-tests/src/option.rs:
```diff
@@ -35,10 +35,16 @@ mod ffi {
+        type OptTestOpaqueRefRustType;
 
         #[swift_bridge(init)]
+        #[swift_bridge(associated_to = OptTestOpaqueRustType)]
```
Need to remove this `associated_to`. I'll open an issue about emitting a compile time error t `#[swift_bridge(init)]` and `#[swift_bridge(associated_to = OptTestOpaqueRustType)]` from both being used.

Removing `associated_to` should fix the problem.

PrismaPhonic replied:
Done - and all passing locally for me now :)

chinedufn replied:
> I'll open an issue

Nevermind the real problem is that if there are two `#[swift_bridge(..)]` attributes the first one gets ignored.

I'll still track it. Thanks.

PrismaPhonic commented:
> Mind pushing new commits instead of force pushing? This way I can see what has changed. I'll squash them all at the end.

Can do - Github in particular has this problem unfortunately. In other platforms you can see a diff between rebases

PrismaPhonic commented:
Addressed all comments - tests are passing for me locally now

chinedufn commented:
Thanks for reporting this issue and thanks for putting together a fix. Appreciate you pushing through the unexpected compile time error after compile time error after compile time error.

Cheers.
